apiserver/facades/controller/firewaller/firewaller.go

```diff
@@ -535,3 +550,164 @@ func (f *FirewallerAPIV5) AreManuallyProvisioned(args params.Entities) (params.B
+	result := params.OpenMachinePortRangesResults{
+		Results: make([]params.OpenMachinePortRangesResult, len(args.Entities)),
+	}
+	canAccess, err := f.accessMachine()
+	if err != nil {
+		return result, err
+	}
+	for i, arg := range args.Entities {
+		machineTag, err := names.ParseMachineTag(arg.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+
+		machine, err := f.getMachine(canAccess, machineTag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+
+		unitPortRanges, err := f.openedPortRangesForOneMachine(machine)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+
+		}
+		result.Results[i].GroupKey = "cidr"
+		result.Results[i].UnitPortRanges = unitPortRanges
+	}
+	return result, nil
+}
+
+func (f *FirewallerAPIV6) openedPortRangesForOneMachine(machine firewall.Machine) ([]params.OpenUnitPortRanges, error) {
+	machPortRanges, err := machine.OpenedPortRanges()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	portRangesByUnit := machPortRanges.ByUnit()
+	if len(portRangesByUnit) == 0 { // no ports open
+		return nil, nil
+	}
+
+	// Look up space to subnet mappings
+	subnetCIDRsBySpaceID, err := machine.SubnetCIDRsBySpaceID()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Fetch application endpoint bindings
+	allApps := set.NewStrings()
+	for unitName := range portRangesByUnit {
+		appName, err := names.UnitApplication(unitName)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		allApps.Add(appName)
+	}
+	allAppBindings, err := f.resolveApplicationBindings(allApps)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Map the port ranges for each unit to one or more subnet CIDRs
+	// depending on the endpoints they apply to.
+	var res []params.OpenUnitPortRanges
+	for unitName, unitPortRanges := range portRangesByUnit {
+		// Already checked for validity; error can be ignored
+		appName, _ := names.UnitApplication(unitName)
+		appBindings := allAppBindings[appName]
+
+		unitRes := params.OpenUnitPortRanges{
+			UnitTag:         names.NewUnitTag(unitName).String(),
+			PortRangeGroups: make(map[string][]params.PortRange),
+		}
+
+		portRangesByCIDR := mapUnitPortsToSubnetCIDRs(unitName, unitPortRanges.ByEndpoint(), appBindings, subnetCIDRsBySpaceID)
+		for cidr, portRanges := range portRangesByCIDR {
+			mappedPorts := make([]params.PortRange, len(portRanges))
+			for i, pr := range portRanges {
+				mappedPorts[i] = params.FromNetworkPortRange(pr)
+			}
+			unitRes.PortRangeGroups[cidr] = mappedPorts
+		}
+
+		res = append(res, unitRes)
+	}
+
+	// Sort by unit tag so we return consistent results
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].UnitTag < res[j].UnitTag
+	})
+
+	return res, nil
+}
+
+func (f *FirewallerAPIV6) resolveApplicationBindings(apps set.Strings) (map[string]map[string]string, error) {
```

Comment:
Given that machines generally don't tend to have to have many applications deployed to them, I thought this call might be good enough in terms of performance (especially on larger models with many deployed applications) as opposed to pulling everything from the DB.

However, if the space topology changes (although I guess the bindings will be the ones changing more often so this might be a moot point), the firewall worker will be making a call for each machine in the model so perhaps we could get all endpoints once at the top of "OpenedMachinePortRanges" and re-use it for each machine. This also would mean that we need to move the client out of that per-machine wrapper (or clone it) so it can pass all machine tags in one go.

Thoughts?